resources/lib/videoplayer.py

```diff
@@ -207,6 +212,59 @@ def _handle_skipping(self):
         utils.crunchy_log("_handle_skipping: starting thread", xbmc.LOGINFO)
         threading.Thread(target=self.thread_check_skipping).start()
 
+    def _handle_upnext(self):
+        try:
+            next_episode = self._stream_data.next_playable_item
+            if not next_episode:
+                utils.crunchy_log("_handle_upnext: No episode or disabled upnext integration")
+                return
+            next_url = view.build_url(
+                {
+                    "series_id": G.args.get_arg("series_id"),
+                    "episode_id": next_episode.episode_id,
+                    "stream_id": next_episode.stream_id
+                },
+                "video_episode_play"
+            )
+            show_next_at_seconds = self._compute_when_episode_ends()
+            # Needs to wait 1s, otherwise, upnext will show next dialog at episode start...
+            xbmc.sleep(1000)
+            utils.crunchy_log("_handle_upnext: Next URL (shown at %ds): %s" % (show_next_at_seconds, next_url))
+            upnext.send_next_info(G.args, self._stream_data.playable_item, next_episode, next_url, show_next_at_seconds, self._stream_data.playable_item_parent)
+        except Exception:
+            utils.crunchy_log("_handle_upnext: Cannot send upnext notification", xbmc.LOGERROR)
+
+    def _compute_when_episode_ends(self) -> int:
+        upnext_mode = G.args.addon.getSetting("upnext_mode")
+        if upnext_mode == "disabled":
+            return None
+
+        video_end = self._stream_data.playable_item.duration
+        fixed_duration = int(G.args.addon.getSetting("upnext_fixed_duration"), 10)
+        result = video_end - fixed_duration
+
+        skip_events_data = self._stream_data.unmodified_skip_events_data
+        if upnext_mode == "fixed" or not skip_events_data or (not skip_events_data.get("credits") and not skip_events_data.get("preview")):
+            return result
+
+        credits_start = skip_events_data.get("credits", {}).get("start")
+        credits_end = skip_events_data.get("credits", {}).get("end")
+        preview_start = skip_events_data.get("preview", {}).get("start")
+        preview_end = skip_events_data.get("preview", {}).get("end")
```

Comment:
> There's another thing: When calculating when to show the UpNext, you refer to the preview skip data, but we don't support/fetch that:
> https://github.com/lumiru/plugin.video.crunchyroll/blob/fd0dd73d8aec9cf35564609986064ab8655a108f/resources/lib/videostream.py#L358

Hum, this is why… Sorry, I have to fix it.

Reply:
You can easily calculate that:
We can only know about a preview if we have info about the credits. So if the end of the credits are less than the total duration (including a small rounding margin of 1 second), there is a preview and it starts when the credits end and ends when the video ends.

Somewhat like this:

```python
if credits_end and video_end - credits_end > 1:
    preview_start = credits_end
    preview_end = video_end
else 
    preview_start = None  # or 0 or whatever. mind that your code for fetching start/end above can return None
    preview_end = None

Reply:
Why could we do it since the data exists in the API? There often are additional scenes after credits, this is why, if data is not present, I consider there is a preview only if difference between video_end and credit_end is less than 20s.

Reply:
However, it cloud be better for readability to apply something similar to your suggest if preview timcaodes are not available.